File: Problems/No.257_Binary_Tree_Paths/binary_tree_paths.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution(object):
    def binaryTreePaths(self, root):
        """
        :type root: TreeNode
        :rtype: List[str]
        """

        # function: searchBTpaths
        # search all Binary Tree paths
        def searchBTpaths(tNode, temp):
            if not tNode:
                return
            else:
                temp += str(tNode.val)
                if ((not tNode.left) and (not tNode.right)):
                    BTPaths.append(temp)
                else:
                    temp += "->"
                    searchBTpaths(tNode.left, temp)
                    searchBTpaths(tNode.right, temp)

        # parameter
        BTPaths = []

        # start computing
        searchBTpaths(root, "")

        return BTPaths

File: Problems/No.257_Binary_Tree_Paths/binary_tree_paths.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:

    // search all Binary Tree paths
    void searchBTpaths(TreeNode* tNode, std::string temp, std::vector<std::string> &PVector) {

        if (tNode == NULL) return;
        else {

            temp += std::to_string(tNode->val);
            if ((tNode->left == NULL) && (tNode->right == NULL)) PVector.push_back(temp);
            else {

                temp += "->";
                searchBTpaths(tNode->left, temp, PVector);
                searchBTpaths(tNode->right, temp, PVector);

            }

        }

    }

    vector<string> binaryTreePaths(TreeNode* root) {
        
        // parameter
        std::vector<std::string> BTPaths;

        // start computing
        searchBTpaths(root, "", BTPaths);

        return BTPaths;

    }
};
